app/ante/cosmos/fees.go

```diff
@@ -0,0 +1,237 @@
+
+package cosmos
+
+import (
+	"fmt"
+	"math"
+
+	errorsmod "cosmossdk.io/errors"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	errortypes "github.com/cosmos/cosmos-sdk/types/errors"
+	authante "github.com/cosmos/cosmos-sdk/x/auth/ante"
+	authtypes "github.com/cosmos/cosmos-sdk/x/auth/types"
+	anteutils "github.com/evmos/evmos/v11/app/ante/utils"
+)
+
+// DeductFeeDecorator deducts fees from the first signer of the tx.
+// If the first signer does not have the funds to pay for the fees,
+// and does not have enough unclaimed staking rewards, then return
+// with InsufficientFunds error.
+// The next AnteHandler is called if fees are successfully deducted.
+//
+// CONTRACT: Tx must implement FeeTx interface to use DeductFeeDecorator
+type DeductFeeDecorator struct {
+	accountKeeper      authante.AccountKeeper
+	bankKeeper         BankKeeper
+	distributionKeeper anteutils.DistributionKeeper
+	feegrantKeeper     authante.FeegrantKeeper
+	stakingKeeper      anteutils.StakingKeeper
+	txFeeChecker       anteutils.TxFeeChecker
+}
+
+// NewDeductFeeDecorator returns a new DeductFeeDecorator.
+func NewDeductFeeDecorator(
+	ak authante.AccountKeeper,
+	bk BankKeeper,
+	dk anteutils.DistributionKeeper,
+	fk authante.FeegrantKeeper,
+	sk anteutils.StakingKeeper,
+	tfc anteutils.TxFeeChecker,
+) DeductFeeDecorator {
+	if tfc == nil {
+		tfc = checkTxFeeWithValidatorMinGasPrices
+	}
+
+	return DeductFeeDecorator{
+		accountKeeper:      ak,
+		bankKeeper:         bk,
+		distributionKeeper: dk,
+		feegrantKeeper:     fk,
+		stakingKeeper:      sk,
+		txFeeChecker:       tfc,
+	}
+}
+
+// AnteHandle ensures that the transaction contains valid fee requirements and tries to deduct those
+// from the account balance or unclaimed staking rewards, which the transaction sender might have.
+func (dfd DeductFeeDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bool, next sdk.AnteHandler) (sdk.Context, error) {
+	feeTx, ok := tx.(sdk.FeeTx)
+	if !ok {
+		return ctx, errorsmod.Wrap(errortypes.ErrTxDecode, "Tx must be a FeeTx")
+	}
+
+	if !simulate && ctx.BlockHeight() > 0 && feeTx.GetGas() <= 0 {
+		return ctx, errorsmod.Wrap(errortypes.ErrInvalidGasLimit, "must provide positive gas")
+	}
+
+	var (
+		priority int64
+		err      error
+	)
+
+	fee := feeTx.GetFee()
+	if !simulate {
+		fee, priority, err = dfd.txFeeChecker(ctx, feeTx)
+		if err != nil {
+			return ctx, err
+		}
+	}
+
+	feePayer := feeTx.FeePayer()
+	feeGranter := feeTx.FeeGranter()
+
+	if err = dfd.deductFee(ctx, tx, fee, feePayer, feeGranter); err != nil {
+		return ctx, err
+	}
+
+	newCtx := ctx.WithPriority(priority)
+
+	return next(newCtx, tx, simulate)
+}
+
+// deductFee checks if the fee payer has enough funds to pay for the fees and deducts them.
+// If the spendable balance is not enough, it tries to claim enough staking rewards to cover the fees.
+func (dfd DeductFeeDecorator) deductFee(ctx sdk.Context, sdkTx sdk.Tx, fees sdk.Coins, feePayer, feeGranter sdk.AccAddress) error {
```

Comment:
this is not fully covered, could we add additional unit tests?

Reply:
additional unit tests added on https://github.com/evmos/evmos/pull/1438